server/grpc_service.go

```diff
@@ -101,6 +114,36 @@ func (s *Server) Tso(stream pdpb.PD_TsoServer) error {
 		if err != nil {
 			return errors.WithStack(err)
 		}
+
+		receiverAddr := request.GetHeader().GetReceiverAddr()
+		if !s.isLocalRequest(receiverAddr) {
+			if redirectStream == nil || lastReceiverAddr != receiverAddr {
+				if cancel != nil {
+					cancel()
+				}
+				client, err := s.getDelegateClient(s.ctx, receiverAddr)
+				if err != nil {
+					return err
+				}
+				redirectStream, cancel, err = s.createRedirectStream(client)
+				if err != nil {
+					return err
+				}
+				lastReceiverAddr = receiverAddr
+			}
+			if err := redirectStream.Send(request); err != nil {
+				return errors.WithStack(err)
+			}
+			resp, err := redirectStream.Recv()
+			if err != nil {
+				return errors.WithStack(err)
+			}
+			if err := stream.Send(resp); err != nil {
+				return errors.WithStack(err)
+			}
```

Comment:
When we meet error here, do we need to clean the `redirectStream` and cancel the `cancel`?

Reply:
When the function returns, defer will cancel the stream.